app/models/manageiq/providers/cloud_manager.rb

```diff
@@ -41,7 +41,7 @@ class << model_name
 
     virtual_has_many :volume_availability_zones, :class_name => "AvailabilityZone", :uses => :availability_zones
 
-    supports_not :auth_key_pair_create
+    delegate_supports :auth_key_pair_create, "AuthKeyPair", :create
```

Comment:
Ergonmically, I wonder if this should be more like:

```ruby
supports :auth_key_pair_create, :delegate => ["AuthKeyPair", :create]
```

What I don't understand in this particular instance is that this was `supports_not` before, but it's `supports` after?  Or does the delegate say it's supported only if the target supports it?

cc @agrare

Reply:
I am not sure if any of this makes sense for this example.  The provider knows if an AuthKey is creatable.  The AuthKey class cannot create a new one, because a provider is needed for that - which does not exist at the AuthKey class level.

Reply:
Usually the subclass has its own `raw_create_*` method so e.g. `ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair.raw_create_key_pair` so if we had `ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair.supports?(:create)` theoretically we wouldn't need to duplicate the same in `ManageIQ::Providers::Amazon::CloudManager.supports?(:auth_key_pair_create)`

As for this previously being supports_not, if the default for AuthKeyPair is `supports_not :create` then we're in the same place.

My question is how does this work with the provider subclasses?  `::AuthKeyPair` will have `supports_not :create` so are we going to have to do the `supports :auth_key_pair_create, :delegate => ["ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair", :create]` or is it going to be "relative" to the calling class.

Reply:
digging deeper, the method signature for `raw_create_key_pair` is `def self.raw_create_key_pair(ext_management_system, create_options)`, which means that it needs to go through the EMS anyway, so I guess the question is should it be an EMS instance method or an AuthKey class method that the caller invokes?

Reply:
> `supports :auth_key_pair_create, :via => ["AuthKeyPair", :create]`

Sure, I can do that (if we want this feature)

> What I don't understand in this particular instance is that this was `supports_not` before, but it's `supports` after? Or does the delegate say it's supported only if the target supports it?

Yes, the core class does not support it. The delegate says to go to the provider specific implementation of the `AuthKeyPair` to determine if `:create` is supported.

> I am not sure if any of this makes sense for this example. The provider knows if an AuthKey is creatable. The AuthKey class cannot create a new one, because a provider is needed for that - which does not exist at the AuthKey class level.

Currently we have:

- [ibm_cloud/cloud_manager.rb](https://github.com/ManageIQ/manageiq-providers-ibm_cloud/blob/9f046aa9e64f5cf84b5fef9c7e08bbfc58495602/app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager.rb#L38)
- [manageiq/cloud_manager.rb](https://github.com/ManageIQ/manageiq/blob/c3e51190374e0494a715ae51dd2c3990539dc354/app/models/manageiq/providers/cloud_manager.rb#L44)
- [amazon/cloud_manager.rb](https://github.com/ManageIQ/manageiq-providers-amazon/blob/0ebe15e51502d6cc7b95019535312a15c30bac66/app/models/manageiq/providers/amazon/cloud_manager.rb#L79)
- [openstack/cloud_manager.rb](https://github.com/ManageIQ/manageiq-providers-openstack/blob/ef393737e21e1f814ee8db8fbf959789186e9860/app/models/manageiq/providers/openstack/cloud_manager.rb#L34)

Looks like I had made a mistake and thought that there was also a `supports :create` defined in [openstack/cloud_manager/auth_key_pair](https://github.com/ManageIQ/manageiq-providers-openstack/blob/ef393737e21e1f814ee8db8fbf959789186e9860/app/models/manageiq/providers/openstack/cloud_manager/auth_key_pair.rb) and the rest of the provider's `AuthKeyPair` implementations.

Maybe we just need a simpler method:

```ruby
class ManageIQ::Providers::CloudManager
  supports :auth_key_pair_create, :if_defined => "AuthKeyPair"
end
```

Or in this case it could just be based upon `AuthKeyPair.class_for_ems(ems)`

```diff
- if ems.supports?(:auth_keypair_create) 
+ unless ManageIQ::Providers::CloudManager::AuthKeyPair.class_for_ems(ems).nil?
```

A, think `class_for_ems` may need to be tweaked (to return nil if it throws a `NameError`) but it could work. I do like the `supports` defining that business logic away.

> Usually the subclass has its own `raw_create_*` method so e.g. `ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair.raw_create_key_pair` so if we had `ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair.supports?(:create)` theoretically we wouldn't need to duplicate the same in `ManageIQ::Providers::Amazon::CloudManager.supports?(:auth_key_pair_create)`

Yes, I'd prefer to only have `AuthKeyPair.supports :create` and not `Ems.supports :auth_key_pair_create`

> My question is how does this work with the provider subclasses? `::AuthKeyPair` will have `supports_not :create` so are we going to have to do the `supports :auth_key_pair_create, :delegate => ["ManageIQ::Providers::Amazon::CloudManager::AuthKeyPair", :create]` or is it going to be "relative" to the calling class.

It was implemented to be relative to the calling class. So these would be identical. (again, showing that I don't mind helpers but I want my base of truth to be based upon crud where possible)

```ruby
ems.supports?(:auth_key_pair_create)
ems.class::AuthKeyPair.supports? :create
```

https://github.com/ManageIQ/manageiq/pull/21473/files#diff-9af00e0aad0105cb96ce15286f7061e24937255556779ea2141a8412c4bc4f5fR229-R257

Reply:
We've had EMS instance methods in the past for things like provider operations and it gets really messy really fast, partially because you have to "namespace" the method names with a prefix.